Cannot parse `..` in slice patterns

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn foo(a: &[u32]) {
    match a {
        [first, ..] => {}
        [.., last] => {}
        _ => {}
    }
}
```

I expected to see this happen: no error in `-fsyntax-only`

Instead, this happened: 
```rust
test.rs:3:17: error: unexpected token ‘..’ in pattern
    3 |         [first, ..] => {}
      |                 ^~
test.rs:3:17: error: failed to parse pattern in slice pattern
test.rs:3:17: error: failed to parse any patterns in match arm
test.rs:3:17: error: failed to parse match arm in match arms
test.rs:2:5: error: failed to parse expr with block in parsing expr statement
    2 |     match a {
      |     ^~~~~
test.rs:2:5: error: failed to parse statement or expression without block in block expression
test.rs:7:1: error: unrecognised token ‘}’ for start of item
    7 | }
      | ^
test.rs:7:1: error: failed to parse item in crate

Analyzing compilation unit

Time variable                                   usr           sys          wall           GGC
 phase setup                        :   0.00 (  0%)   0.00 (  0%)   0.01 (100%)   137k ( 95%)
 TOTAL                              :   0.00          0.00          0.01          145k
Extra diagnostic checks enabled; compiler may run slowly.
Configure with --enable-checking=release to disable checks.
```

### Meta
0d5fe7b9aee

--- a/gcc/rust/ast/rust-pattern.h
+++ b/gcc/rust/ast/rust-pattern.h
@@ -194,15 +194,26 @@ protected:
 class RestPattern : public Pattern
 {
   Location locus;
+  NodeId node_id;
 
 public:
   std::string as_string () const override { return ".."; }
 
-  RestPattern (Location locus) : locus (locus) {}
+  RestPattern (Location locus)
+    : locus (locus), node_id (Analysis::Mappings::get ()->get_next_node_id ())
+  {}
 
   Location get_locus () const override final { return locus; }
 
   void accept_vis (ASTVisitor &vis) override;
+
+  NodeId get_pattern_node_id () const override final { return node_id; }
+
+protected:
+  RestPattern *clone_pattern_impl () const override
+  {
+    return new RestPattern (*this);
+  }
 };
 
 // Base range pattern bound (lower or upper limit) - abstract

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -10737,6 +10737,10 @@ Parser<ManagedTokenSource>::parse_pattern_no_alt ()
       lexer.skip_token ();
       return std::unique_ptr<AST::WildcardPattern> (
 	new AST::WildcardPattern (t->get_locus ()));
+    case DOT_DOT:
+      lexer.skip_token ();
+      return std::unique_ptr<AST::RestPattern> (
+	new AST::RestPattern (t->get_locus ()));
     case REF:
     case MUT:
       return parse_identifier_pattern ();

--- a/gcc/testsuite/rust/compile/slice_rest_pattern.rs
+++ b/gcc/testsuite/rust/compile/slice_rest_pattern.rs
@@ -1,0 +1,8 @@
+// { dg-options "-fsyntax-only" }
+fn foo(a: &[u32]) {
+    match a {
+        [first, ..] => {}
+        [.., last] => {}
+        _ => {}
+    }
+}

